### DISCUSSION TITLE
HiGHS benchmark for PyPSA problems and solver options

### DISCUSSION THREAD

**Discussion by pz-max** [ERGO-Code/HiGHS#1185] [Q&A]
Hi HiGHS team, @fneum was playing a little bit around with HiGHS and created a benchmark that is updating every 5 days, is automatically looking for the newest HiGHS version (now 1.5.0dev0, 27.02.23) and is comparing it performance with the 1.1.2.dev3 version. The benchmark is [hosted](https://fneum.github.io/highs-pypsa-progress/) on a GitHub pages (prototype), with the code being available on [GitHub here](https://github.com/fneum/highs-pypsa-progress). One goal is to extend this benchmark and test various energy system modelling specific problems with HiGHS and potentially compare them to other open-source alternatives e.g. cbc, glpk, scip.

The benchmark is currently using the Python **Linopy** interface for HiGHS (thanks to dev. @FabianHofmann) that can:
- Pass lp/mps files to HiGHS with this [function](https://github.com/PyPSA/linopy/blob/5b7fe29bc99261751886957b347cebadb1ee2a54/linopy/solvers.py#L304-L340).
- Or write them "directly" with highspy without intermediary lp/mps files using [this function](https://github.com/PyPSA/linopy/blob/5b7fe29bc99261751886957b347cebadb1ee2a54/linopy/io.py#L276-L317). The direct pass to highspy makes it super performant. However, the current implementation does not track Linopy variable and constraint labels which probably requires a custom interface as noted by @FabianHofmann)

For the benchmark, we currently use the [HiGHS default options.](https://github.com/PyPSA/linopy/blob/5b7fe29bc99261751886957b347cebadb1ee2a54/linopy/solvers.py#L324-L329). Let us know if we should pass any other option that we could add [here](https://github.com/fneum/highs-pypsa-progress/blob/014ac346853578e0d4d5797744fc1ad522901f8f/scripts/benchmark.py#L9-L16).

- All HiGHS options are here in code: https://github.com/ERGO-Code/HiGHS/blob/ce351010ec7be8b7aae73d847d02669c927c53da/src/lp_data/HighsOptions.h#L252-L265
- Also I noted that the documentation looks quite cool now! Is there a way to edit the [documentation](https://ergo-code.github.io/HiGHS/options.html)? 

Thanks for all the fascinating work! :1st_place_medal:

**Comment by jajhall** [comment]
I'm puzzled as to why you can't pass Linopy variable and constraint labels as column and row names and think that you need a custom interface 

The only option that you might try is run_crossover=off since I recall that you don't need a basic solution.

At the moment the documentation is WIP, so is not editable. Once it's complete, edits could be suggested by checking out a branch of HiGHS `latest`, editing the `.md` files, and making a PR

Sorry not to reply sooner, but I've only just switched on notifications for discussions

**Reply by pz-max** [reply]
- @FabianHofmann might respond in the next days on the Linopy stuff
- Thanks for the recommendation. Indeed crossover can be turned off
- Good prospect on the documentation :100: 

FYI, the [benchmark](https://fneum.github.io/highs-pypsa-progress/) is not working correctly at the moment. Fabian N. noted that one and the same HiGHS version is downloaded which is not really the purpose of comparing different versions... We need to fix that :+1:

**Reply by FabianHofmann** [reply]
Hey all, thanks @jajhall for the reply. Just to correct here, the label tracking in linopy actually works. We had some issues with using custom labels for variables and constraints with highspy in the beginning but that was solved some minor versions ago :)
